reaper/reaper.go

```diff
@@ -47,60 +49,54 @@ func Ready() {
 
 // Reaper finds resources and deals with them
 type Reaper struct {
-	stopCh chan bool
+	*cron.Cron
 }
 
 // NewReaper is a Reaper constructor shorthand
 func NewReaper() *Reaper {
 	return &Reaper{
-		stopCh: make(chan bool),
+		Cron: cron.New(),
 	}
 }
 
-// Start begins Reaper execution in a new goroutine
+// Start begins Reaper's schedule
 func (r *Reaper) Start() {
-	go r.start()
-}
+	// adding as a job runs r.Run() every interval
+	r.Cron.Schedule(cron.Every(config.Notifications.Interval.Duration), r)
+	r.Cron.Start()
 
-// Stop closes a Reaper's stop channel
-func (r *Reaper) Stop() {
-	close(r.stopCh)
-	for _, e := range *events {
-		if err := e.Cleanup(); err != nil {
-			log.Error(err.Error())
-		}
-	}
-}
+	// initial run
+	go r.Run()
 
-// unexported start is continuous loop that reaps every
-// time interval
-func (r *Reaper) start() {
 	// if loading from saved state file (overriding AWS states)
 	if config.LoadFromStateFile {
 		r.LoadState(config.StateFile)
 	}
+}
 
-	// make a list of all eligible instances
-	for {
-		r.Once()
-		select {
-		case <-time.After(config.Notifications.Interval.Duration):
-		case <-r.stopCh: // time to exit!
-			log.Debug("Stopping reaper on stop channel message")
-			return
+// Stop stops Reaper's schedule
+func (r *Reaper) Stop() {
+	log.Debug("Stopping Reaper")
+	for _, e := range *events {
+		if err := e.Cleanup(); err != nil {
+			log.Error(err.Error())
 		}
 	}
+	r.Cron.Stop()
 }
 
-// Once is run once every time interval by start
-// it is intended to handle all reaping logic
-func (r *Reaper) Once() {
+// Run handles all reaping logic
+// conforms to the cron.Job interface
+func (r *Reaper) Run() {
+	schedule = cron.New()
+	schedule.Start()
```

Comment:
By default it is empty. Starting it simply sleeps until a job is added, at
which point it wakes momentarily, checks the next time it should wake up
(based on the time the new job should run), and then sleeps.

See: http://godoc.org/github.com/robfig/cron
Quoted below:
Implementation

Cron entries are stored in an array, sorted by their next activation time.
Cron sleeps until the next job is due to be run.

Upon waking:

- it runs each entry that is active on that second
- it calculates the next run times for the jobs that were run
- it re-sorts the array of entries by next activation time.
- it goes to sleep until the soonest job.



On Mon, Jul 27, 2015 at 2:11 PM, whd <notifications@github.com> wrote:

> In reaper/reaper.go
> <https://github.com/mozilla-services/reaper/pull/51#discussion_r35587653>:
>
> >  }
> >
> > -// Once is run once every time interval by start
> > -// it is intended to handle all reaping logic
> > -func (r *Reaper) Once() {
> > +// Run handles all reaping logic
> > +// conforms to the cron.Job interface
> > +func (r *Reaper) Run() {
> > +	schedule = cron.New()
> > +	schedule.Start()
>
> What is a new cron schedule initialized to by default?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mozilla-services/reaper/pull/51/files#r35587653>.
>